src/components/ResourceList/components/Item/Item.tsx

```diff
@@ -65,273 +64,258 @@ export type CombinedProps =
+    selectMode && styles.selectMode,
+    persistActions && styles.persistActions,
+    focusedInner && styles.focusedInner,
+  );
+
+  let actionsMarkup: React.ReactNode | null = null;
+  let disclosureMarkup: React.ReactNode | null = null;
+
+  if (shortcutActions && !loading) {
+    if (persistActions) {
+      actionsMarkup = (
+        <div className={styles.Actions} onClick={stopPropagation}>
+          <ButtonGroup>
+            {buttonsFrom(shortcutActions, {
+              size: 'slim',
+              plain: true,
+            })}
+          </ButtonGroup>
+        </div>
+      );
 
-  private handleAnchorFocus = () => {
-    this.setState({focused: true, focusedInner: false});
-  };
+      disclosureMarkup = (
+        <div className={styles.Disclosure} onClick={stopPropagation}>
+          <Popover
+            activator={
+              <Button
+                aria-label={intl.translate(
+                  'Polaris.ResourceList.Item.actionsDropdown',
+                )}
+                onClick={handleActionsClick}
+                plain
+                icon="horizontalDots"
+              />
+            }
+            onClose={handleCloseRequest}
+            active={actionsMenuVisible}
+          >
+            <ActionList items={shortcutActions} />
+          </Popover>
+        </div>
+      );
+    } else {
+      actionsMarkup = (
+        <div className={styles.Actions} onClick={stopPropagation}>
+          <ButtonGroup segmented testID="ShortcutActions">
+            {buttonsFrom(shortcutActions, {
+              size: 'slim',
+            })}
+          </ButtonGroup>
+        </div>
+      );
+    }
+  }
+
+  const content = children ? (
+    <div className={styles.Content}>{children}</div>
+  ) : null;
+
+  const containerMarkup = (
+    <div testID="Item-Content" className={styles.Container} id={id}>
+      {ownedMarkup}
+      {content}
+      {actionsMarkup}
+      {disclosureMarkup}
+    </div>
+  );
+
+  const tabIndex = loading ? -1 : 0;
+
+  const accessibleMarkup = url ? (
+    <UnstyledLink
+      aria-describedby={id}
+      aria-label={accessibilityLabel}
+      className={styles.Link}
+      url={url}
+      onFocus={handleAnchorFocus}
+      onBlur={handleFocusedBlur}
+      tabIndex={tabIndex}
+    />
+  ) : (
+    <button
+      className={styles.Button}
+      aria-label={accessibilityLabel}
+      aria-controls={ariaControls}
+      aria-expanded={ariaExpanded}
+      onClick={handleClick}
+      onFocus={handleAnchorFocus}
+      onBlur={handleFocusedBlur}
+      tabIndex={tabIndex}
+    />
+  );
+
+  return (
+    <div
+      ref={node}
+      className={className}
+      onClick={handleClick}
+      onFocus={handleFocus}
+      onBlur={handleBlur}
+      onMouseDown={handleMouseDown}
+      onKeyUp={handleKeypress}
+      testID="Item-Wrapper"
+      data-href={url}
+    >
+      {accessibleMarkup}
+      {containerMarkup}
+    </div>
+  );
+
+  function handleAnchorFocus() {
```

Comment:
Having a bunch of functions defined beneath the return feels odd to me. Can you pull these up to be towards the start of the function.

This is a functional component and callback functions that we define need to be wrapped in [useCallback](https://reactjs.org/docs/hooks-reference.html#usecallback). Otherwise a new callback function ends up being created every render, and when you pass that down to a child component you end up rerendering it - as its props have changed becase it's onWhatever handler is a different function to the one that was used before.

Wrapping these declarations in useCallback means the same function will be passed to the child component if given the same arguments, thus avoiding unneeded rerenders.

```js
const handleFocusedBlur  = useCallback(() => {
   setFocusState({focused: true, focusedInner: true});
  }, [setFocusState]);
```

See https://github.com/BPScott/ksw-quiz/commit/30b4c3f609f6eeae8edf550e466af129e785bf87 for an example of how I dealt with this in my little toy project for playing with react

Reply:
I find putting the functions below cleans up the code, Tim does the same as well I think. However, if we use functional expressions we'll want to pull it up. Thanks for pointing this out, I was worried about it but wasn't aware there was a hook to solve it 🙏